### PR TITLE
i.MX support #3

### PR DESCRIPTION
**Pull request by clementfaure**
Hello,

This PR adds the support for the imx7ulpevk platform.
It also fixes few issues and includes a re-work of SoC detection functions.

Thank you!

### CODE REVIEW

**Comment by clementfaure** [comment]
@MrVan Could you take time to review it?

File: core/arch/arm/plat-imx/imx-common.c [hunk 1]
```diff
@@ -13,34 +13,43 @@
...
-#if defined(CFG_MX7)
-		reg = io_read32(anatop_addr + OFFSET_DIGPROG_IMX7D);
-#elif defined(CFG_MX6SL)
-		reg = io_read32(anatop_addr + OFFSET_DIGPROG_IMX6SL);
-#else
-		reg = io_read32(anatop_addr + OFFSET_DIGPROG);
-#endif
-	}
+	/* Set the CPU type */
+	imx_cpu_type = ((digprog >> 16) & 0xFF);
```
Comment (etienne-lms) [inline core/arch/arm/plat-imx/imx-common.c]: can drop outer parantheses

File: core/arch/arm/plat-imx/imx-common.c [hunk 2]
```diff
@@ -13,34 +13,43 @@
...
 
-	return reg;
+	/* Set the SOC revision = (Major + 1).(Minor) */
+	imx_soc_revision =
+		(((digprog & 0xFF00) >> 4) + 0x10) | (digprog & 0x0F);
```
Comment (etienne-lms) [inline core/arch/arm/plat-imx/imx-common.c]: suggestion: may prefer macros with a friendly name than magic values

File: core/arch/arm/plat-imx/imx-common.c [hunk 3]
```diff
@@ -13,34 +13,43 @@
 #include <mm/core_memprot.h>
 #include <platform_config.h>
 
-static uint32_t imx_digproc(void)
+static int imx_cpu_type = (-1);
+static int imx_soc_revision = (-1);
```
Comment (etienne-lms) [inline core/arch/arm/plat-imx/imx-common.c]: this mandates major < 0x80.
minor: drop parentheses

File: core/arch/arm/plat-imx/imx-common.c [hunk 4]
```diff
@@ -13,34 +13,43 @@
...
-	return reg;
+	/* Set the SOC revision = (Major + 1).(Minor) */
+	imx_soc_revision =
+		(((digprog & 0xFF00) >> 4) + 0x10) | (digprog & 0x0F);
 }
 
 static uint32_t imx_soc_rev_major(void)
 {
-	return ((imx_digproc() & 0xff00) >> 8) + 1;
+	if (imx_soc_revision < 0)
```
Comment (etienne-lms) [inline core/arch/arm/plat-imx/imx-common.c]: these test mandates major < 0x80. Maybe add an inline comment for this.

File: core/arch/arm/plat-imx/imx-common.c [hunk 5]
```diff
@@ -19,15 +19,18 @@ static int imx_soc_revision = (-1);
 static void imx_digproc(void)
 {
 	uint32_t digprog;
-	vaddr_t anatop_addr;
+	vaddr_t __maybe_unused anatop_addr;
```
Comment (etienne-lms) [inline core/arch/arm/plat-imx/imx-common.c]: by the way, could you default init these 2 to something (i.e.`0`)

File: core/arch/arm/plat-imx/pm/psci.c [hunk 6]
```diff
@@ -66,20 +66,18 @@ int psci_cpu_on(uint32_t core_idx, uint32_t entry,
...
-		io_write32(va + SRC_GPR1_MX7 + core_idx * 8, val);
-
-		imx_gpcv2_set_core1_pup_by_software();
 
-		/* release secondary core */
-		val = io_read32(va + SRC_A7RCR1);
-		val |=  BIT32(SRC_A7RCR1_A7_CORE1_ENABLE_OFFSET +
-			      (core_idx - 1));
-		io_write32(va + SRC_A7RCR1, val);
+#ifdef CFG_MX7
```
Comment (MrVan) [inline core/arch/arm/plat-imx/pm/psci.c]: Why change to ifdef?

File: core/arch/arm/plat-imx/registers/imx7ulp.h [hunk 7]
```diff
@@ -0,0 +1,53 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright 2017-2018 NXP
```
Comment (MrVan) [inline core/arch/arm/plat-imx/registers/imx7ulp.h]: 2019 should be included.

File: core/drivers/imx_lpuart.c [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+{
+	struct imx_uart_data *pd =
+		container_of(chip, struct imx_uart_data, chip);
+
+	return io_pa_or_va(&pd->base);
+}
+
+static void imx_lpuart_flush(struct serial_chip *chip __unused)
+{
+
```
Comment (MrVan) [inline core/drivers/imx_lpuart.c]: Drop empty line.

File: core/drivers/imx_lpuart.c [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+	return ch;
+}
+
+static void imx_lpuart_putc(struct serial_chip *chip, int ch)
+{
+	vaddr_t base = chip_to_base(chip);
+
+	while (!(io_read32(base + STAT) & STAT_TDRE))
+		__asm__ __volatile__("" : : : "memory");
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/drivers/imx_lpuart.c]: @MrVan Is this memory barrier needed?

File: core/arch/arm/plat-imx/pm/psci.c (hunk 6 above)
Comment (clementfaure) [reply core/arch/arm/plat-imx/pm/psci.c]: gpcv2 functions are compiled only for mx7s/d

File: core/arch/arm/plat-imx/imx-common.c (hunk 4 above)
Comment (clementfaure) [reply core/arch/arm/plat-imx/imx-common.c]: Hi @etienne-lms , I don't get what you want me to comment here.

Comment (etienne-lms) [reply core/arch/arm/plat-imx/imx-common.c]: sorry, it's a 32bit, large enough for the stored value.
Discard my comment :)

Comment (clementfaure) [reply core/arch/arm/plat-imx/imx-common.c]: Ok :)

File: core/arch/arm/plat-imx/imx-common.c (hunk 1 above)
Comment (clementfaure) [reply core/arch/arm/plat-imx/imx-common.c]: Addressed in a5496ac

File: core/arch/arm/plat-imx/imx-common.c (hunk 2 above)
Comment (clementfaure) [reply core/arch/arm/plat-imx/imx-common.c]: Addressed in a5496ac

File: core/arch/arm/plat-imx/imx-common.c (hunk 3 above)
Comment (clementfaure) [reply core/arch/arm/plat-imx/imx-common.c]: Addressed in a5496ac

File: core/drivers/imx_lpuart.c [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+static void imx_lpuart_flush(struct serial_chip *chip __unused)
+{
+
+}
+
+static int imx_lpuart_getchar(struct serial_chip *chip)
+{
+	int ch;
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/drivers/imx_lpuart.c]: Addressed in a5496ac

File: core/arch/arm/plat-imx/imx-common.c (hunk 5 above)
Comment (clementfaure) [reply core/arch/arm/plat-imx/imx-common.c]: Addressed in a5496ac

File: core/arch/arm/plat-imx/registers/imx7ulp.h (hunk 7 above)
Comment (clementfaure) [reply core/arch/arm/plat-imx/registers/imx7ulp.h]: Addressed in a5496ac

File: core/drivers/imx_lpuart.c (hunk 8 above)
Comment (clementfaure) [reply core/drivers/imx_lpuart.c]: Addressed in a5496ac

File: core/arch/arm/plat-imx/imx-common.c [hunk 11]
```diff
@@ -13,34 +13,54 @@
...
+#endif
 }
 
 static uint32_t imx_soc_rev_major(void)
 {
-	return ((imx_digproc() & 0xff00) >> 8) + 1;
+	if (imx_soc_revision < 0)
+		imx_digproc();
+
+	return (imx_soc_revision >> 4);
```
Comment (etienne-lms) [inline core/arch/arm/plat-imx/imx-common.c]: minor: can drop outer parentheses.

File: core/arch/arm/plat-imx/registers/imx7ulp.h [hunk 12]
```diff
@@ -0,0 +1,53 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright 2017-2019 NXP
+ *
```
Comment (etienne-lms) [inline core/arch/arm/plat-imx/registers/imx7ulp.h]: can remove empty line

File: core/drivers/imx_lpuart.c [hunk 13]
```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2017-2019 NXP
+ *
```
Comment (etienne-lms) [inline core/drivers/imx_lpuart.c]: can remove empty line

File: core/drivers/imx_lpuart.c [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+static int imx_lpuart_getchar(struct serial_chip *chip)
+{
+	int ch = 0;
+	vaddr_t base = chip_to_base(chip);
+
+	while (io_read32(base + STAT) & STAT_RDRF)
+		;
+
+	ch = (io_read32(base + DATA) & 0x3ff);
```
Comment (etienne-lms) [inline core/drivers/imx_lpuart.c]: can drop outer parentheses

File: core/drivers/imx_lpuart.c [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+};
+KEEP_PAGER(imx_lpuart_ops);
+
+void imx_uart_init(struct imx_uart_data *pd, paddr_t base)
+{
+	pd->base.pa = base;
+	pd->chip.ops = &imx_lpuart_ops;
+
+	/*
+	 * Do nothing, debug uart(sc lpuart) share with normal world,
```
Comment (etienne-lms) [inline core/drivers/imx_lpuart.c]: s/share/shared/

File: core/include/drivers/imx_wdog.h [hunk 16]
```diff
@@ -40,6 +40,25 @@
 #define WDT_SEQ1	0x5555
 #define WDT_SEQ2	0xAAAA
 
+/* 7ULP */
+#define WDOG_CNT	0x4
+#define WDOG_TOVAL	0x8
+
+#define REFRESH_SEQ0	0xA602
+#define REFRESH_SEQ1	0xB480
+#define REFRESH		((REFRESH_SEQ1 << 16) | (REFRESH_SEQ0))
```
Comment (etienne-lms) [inline core/include/drivers/imx_wdog.h]: s/(REFRESH_SEQ0)/REFRESH_SEQ0/

File: core/include/drivers/imx_wdog.h [hunk 17]
```diff
@@ -40,6 +40,25 @@
...
+
+#define UNLOCK_SEQ0	0xC520
+#define UNLOCK_SEQ1	0xD928
+#define UNLOCK		((UNLOCK_SEQ1 << 16) | (UNLOCK_SEQ0))
```
Comment (etienne-lms) [inline core/include/drivers/imx_wdog.h]: ditto

File: core/arch/arm/plat-imx/imx-common.c (hunk 11 above)
Comment (clementfaure) [reply core/arch/arm/plat-imx/imx-common.c]: Addressed aff7581

File: core/arch/arm/plat-imx/registers/imx7ulp.h (hunk 12 above)
Comment (clementfaure) [reply core/arch/arm/plat-imx/registers/imx7ulp.h]: aff7581

File: core/drivers/imx_lpuart.c (hunk 13 above)
Comment (clementfaure) [reply core/drivers/imx_lpuart.c]: aff7581

File: core/drivers/imx_lpuart.c (hunk 14 above)
Comment (clementfaure) [reply core/drivers/imx_lpuart.c]: Addressed in aff7581

File: core/drivers/imx_lpuart.c (hunk 15 above)
Comment (clementfaure) [reply core/drivers/imx_lpuart.c]: Addressed in aff7581

File: core/include/drivers/imx_wdog.h (hunk 16 above)
Comment (clementfaure) [reply core/include/drivers/imx_wdog.h]: Addressed in aff7581

File: core/include/drivers/imx_wdog.h (hunk 17 above)
Comment (clementfaure) [reply core/include/drivers/imx_wdog.h]: Addressed in aff7581

File: core/arch/arm/plat-imx/registers/imx7ulp.h (hunk 12 above)
Comment (clementfaure) [reply core/arch/arm/plat-imx/registers/imx7ulp.h]: Addressed aff7581

File: core/drivers/imx_lpuart.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (MrVan) [reply core/drivers/imx_lpuart.c]: @clementfaure It could be dropped.

**Comment by MrVan** [comment]
With the minor fix in lpuart, you could add `Reviewed-by: Peng Fan <peng.fan@nxp.com>` for 1,2,3,6,7.

**Comment by clementfaure** [comment]
@etienne-lms Minor fixes squashed.
@MrVan lpuart fixed and tag applied.

File: core/drivers/imx_lpuart.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/drivers/imx_lpuart.c]: Got it!

File: core/arch/arm/plat-imx/imx-common.c [hunk 18]
```diff
@@ -35,9 +35,13 @@ static void imx_digproc(void)
 	/* Set the CPU type */
 	imx_cpu_type = CPU_TYPE(digprog);
 
+#ifdef CFG_MX7
+	imx_soc_revision = digprog & 0xFF;
+#else
 	/* Set the SOC revision: = (Major + 1)[11:4] | (Minor[3:0]) */
 	imx_soc_revision =
 		(SOC_REV_MAJOR(digprog) << 4) | SOC_REV_MINOR(digprog);
+#endif
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/arch/arm/plat-imx/imx-common.c]: It's true that doing so looks cleaner. However, in my next PR, I will add support for i.MX8Q and these boards require a similar "workaround" and I don't want to implement imx_digproc() for each board. Here is a preview of the function:

```
static void imx_digproc(void)
{
	uint32_t digprog = 0;
	vaddr_t __maybe_unused anatop_addr = 0;

#if defined(CFG_MX7ULP)
	digprog = SOC_MX7ULP << 16;
#elif defined(CFG_IMX8QX)
	digprog = SOC_MX8QX << 16;
#elif defined(CFG_IMX8QM)
	digprog = SOC_MX8QM << 16;
#else
	anatop_addr = core_mmu_get_va(ANATOP_BASE, MEM_AREA_IO_SEC);

	if (!anatop_addr)
		return;

	digprog = io_read32(anatop_addr + DIGPROG_OFFSET);
#endif
	/* Set the CPU type */
	imx_cpu_type = CPU_TYPE(digprog);

#ifdef CFG_MX7
	imx_soc_revision = digprog & 0xFF;
#else
	/* Set the SOC revision: = (Major + 1)[11:4] | (Minor[3:0]) */
	imx_soc_revision =
		(SOC_REV_MAJOR(digprog) << 4) | SOC_REV_MINOR(digprog);
#endif
}
```
Thank you for your suggestion :)

[the remark this answers is not in the document]
Comment (clementfaure) [reply core/arch/arm/plat-imx/imx-common.c]: @etienne-lms Tags applied!

**Pull request merged**